---
title: "Comparison of effect estimates from CHIP gene burden by case type"
author: "Austin Hammermeister Suger"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE,
                      comment = "", fig.path = 'figures/forest/', dev='jpeg',
                      cache = FALSE)
knitr::opts_knit$set(eval.after = "fig.cap")
library(pacman)

p_load(rlang, tidyverse, data.table,
       reactable, R.utils,
       ggplot2, ggrepel,
       fastqq, stringi, topr,
       RColorBrewer, cowplot,
       ggtext)

```

```{r data}
CHIP_gene_results_UKB = CHIP_gene_results_AoU = NULL 
for (i in c(2, 4, 9, 17, 20)) {
  UKB_chr_data = fread(paste0("/lindstroem/austin_working/Dissertation/UKBB/gene_based_results/individual_cancer_CHIP_gene_check/chr",i,"_CHIP_gene_count_results.txt")) 
  AoU_chr_data = fread(paste0("/lindstroem/austin_working/Dissertation/AoU/gene_based_results/individual_cancers_CHIP_gene_check/chr",i,"_CHIP_gene_count_results.txt"))
  CHIP_gene_results_UKB = rbind.data.frame(CHIP_gene_results_UKB,UKB_chr_data)
  CHIP_gene_results_AoU = rbind.data.frame(CHIP_gene_results_UKB,AoU_chr_data)
}
# Columns to exclude from renaming
exclude_cols <- c("Cancer", "Gene", "Group")

CHIP_gene_results_UKB <- CHIP_gene_results_UKB %>%
  rename_with(.fn = ~ paste0(., "_UKB"), .cols = -all_of(exclude_cols))

CHIP_gene_results_AoU <- CHIP_gene_results_AoU %>%
  rename_with(.fn = ~ paste0(., "_AoU"), .cols = -all_of(exclude_cols))

CHIP_gene_results = merge(CHIP_gene_results_UKB, CHIP_gene_results_AoU, by = exclude_cols)
```


```{r plot}
CHIP_gene_results <- CHIP_gene_results %>%
  group_by(Group) %>%
  mutate(
    outlier = (abs(PrevalentCase_coef_UKB - mean(PrevalentCase_coef_UKB, na.rm = TRUE)) > 2 * sd(PrevalentCase_coef_UKB, na.rm = TRUE) & PrevalentCase_p_UKB < 0.05) |
              (abs(PrevalentCase_coef_AoU - mean(PrevalentCase_coef_AoU, na.rm = TRUE)) > 2 * sd(PrevalentCase_coef_AoU, na.rm = TRUE) & PrevalentCase_p_AoU < 0.05),
    label = ifelse(outlier, paste(Cancer, Gene, sep = "_"), NA)
  ) %>%
  ungroup()  # Ungroup for plotting

plot = ggplot(CHIP_gene_results, aes(x = PrevalentCase_coef_UKB, y = PrevalentCase_coef_AoU)) +
  geom_point(aes(color = outlier), size = 3, shape = 1) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
  geom_text_repel(aes(label = label), size = 3, na.rm = TRUE) +
  scale_color_manual(values = c("FALSE" = "black", "TRUE" = "red")) +
  facet_wrap(~ Group, scales = "free") +
  theme_classic(base_size = 16) +
  theme(legend.position = "bottom") +
  labs(color = "Outlier",
       x = "mean effect of prevalent case status on rare variant burden in UKB",
       y = "mean effect of prevalent case status on rare variant burden in AoU")


ggsave(
  "./figures/CHIP_gene_lm_results.jpeg",
  plot,
  width = 12, height = 12, dpi = 600
)

```
